website/blog/2024/11-15-graphql-persisted-query/index.md

```diff
@@ -0,0 +1,90 @@
+slug: graphql-persisted-query
+title: 'Reverse engineering GraphQL persistedQuery extension'
+description: 'Learn how to do reverse engineering on persistedQuery extension by GraphQL and reveal the query hash needed for scraping.'
+image: https://raw.githubusercontent.com/souravjain540/crawlee-first-blog/main/og-image.webp
+authors: [SauravJ, MatejV]
+---
+
+GraphQL is a query language for getting deeply nested structured data from a website's backend, similar to MongoDB queries.
+
+The request is usually a POST to some general `/graphql` endpoint with a body like this:
+
+![GraphQL Query](./img/graphql.webp)
+
+When scraping data from websites using GraphQL, it’s common to inspect the network requests in developer tools to find the exact queries being used. However, on some websites, you might notice that the GraphQL query itself isn’t visible in the request. Instead, you only see a cryptic hash value. This can be confusing and makes it harder to understand how data is being requested from the server.
+
+This is because some websites use a feature called ["persisted queries.](https://www.apollographql.com/docs/apollo-server/performance/apq/) It's a performance optimization that reduces the amount of data sent with each request by replacing the full query text with a precomputed hash. While this improves website speed and efficiency, it introduces challenges for scraping because the query text isn’t readily available.
+
+
+![Persisted Query Reverse Engineering](./img/graphql-persisted-query.webp)
+
+<!-- truncate -->
+
+TLDR: the client computes the sha256 hash of the `query` text and only sends that hash. In addition, you can possibly fit all of this into the query string of a GET request, making it easily cachable. Below is an example request from Zillow
+
+![Request from Zillow](./img/zillow.webp)
+
+As you can see, it’s just some metadata about the persistedQuery extension, the hash of the query, and variables to be embedded in the query.
+
+Here’s another request from expedia.com, sent as a POST, but with the same extension:
+
+This primarily optimizes website performance, but it creates several challenges for web scraping:
+
+- GET requests are usually more prone to being blocked.
+- Hidden Query Parameters: We don’t know the full query, so if the website responds with a “Persisted query not found” error (asking us to send the query in full, not just the hash), we can’t send it.
+- Once the website changes even a little bit and the clients start asking for a new query - even though the old one might still work, the server will very soon forget its ID/hash, and your request with this hash will never work again, since you can’t “remind” the server of the full query text.
+
+For various reasons, you might need to extract the entire GraphQL query text, but this can be tricky. While you could inspect the website’s JavaScript to find the query text, it’s often dynamically constructed from multiple fragments, making it hard to piece together.
```

Comment:
I mean, the reason is that we want to call the GraphQL API directly so that we don't have to scrape the HTML, right? Or am I missing something?